binding.c

```diff
@@ -92,40 +98,93 @@ parse_address (struct sockaddr *name, char *ip, size_t size, int *port, int *fam
   }
 }
 
+/**
+ * When a native function is executing on uv event;
+ * (cannot be surrounded by try-catch block)
+ *
+ * redirect pending exceptions to global uncaught handler
+ * to avoid process abort (given an uncaught handler is installed).
+ */
+static inline void
+redirect_to_uncaught (js_env_t *env, int err) {
+  if (err == 0) return;
+  assert(err == js_pending_exception || err == js_uncaught_exception);
+
+  bool is_pending;
+  err = js_is_exception_pending(env, &is_pending);
+  assert(err == 0);
+
+  if (is_pending) {
+    js_value_t *fatal_exception;
+    err = js_get_and_clear_last_exception(env, &fatal_exception);
+    assert(err == 0);
+
+    err = js_fatal_exception(env, fatal_exception);
+    assert(err == 0);
+  }
+}
+
+static inline int
+get_buffer_info (js_env_t *env, js_value_t *buffer, void **data, size_t *len) {
+#ifdef CHECKED_GET_BUFFER_INFO
+  int err;
+  bool is_typedarray;
+  err = js_is_typedarray(env, buffer, &is_typedarray);
+  if (err) return err;
+  if (is_typedarray) return js_get_typedarray_info(env, buffer, NULL, data, len, NULL, NULL);
+
+  bool is_arraybuffer;
+  err = js_is_arraybuffer(env, buffer, &is_arraybuffer);
+  if (err) return err;
+  if (is_arraybuffer) return js_get_arraybuffer_info(env, buffer, data, len);
+
+  return -1;
+#else
+  // all buffers allocated through b4a are typed-arrays
+  return js_get_typedarray_info(env, buffer, NULL, data, len, NULL, NULL);
+#endif
+}
+
+
 static void
 on_udx_send (udx_socket_send_t *req, int status) {
+  int err;
   udx_napi_socket_t *n = (udx_napi_socket_t *) req->socket;
   if (n->udx->exiting) return;
 
-  napi_env env = n->env;
+  js_env_t *env = n->env;
 
-  napi_handle_scope scope;
-  napi_open_handle_scope(env, &scope);
+  js_handle_scope_t *scope;
+  err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
 
-  napi_value ctx;
-  napi_get_reference_value(env, n->ctx, &ctx);
+  js_value_t *ctx;
+  err = js_get_reference_value(env, n->ctx, &ctx);
+  assert(err == 0);
 
-  napi_value callback;
-  napi_get_reference_value(env, n->on_send, &callback);
+  js_value_t *callback;
+  err = js_get_reference_value(env, n->on_send, &callback);
+  assert(err == 0);
 
-  napi_value argv[2];
-  napi_create_int32(env, (uintptr_t) req->data, &(argv[0]));
-  napi_create_int32(env, status, &(argv[1]));
+  js_value_t *argv[2];
+  err = js_create_int32(env, (uintptr_t) req->data, &(argv[0]));
+  assert(err == 0);
+  err = js_create_int32(env, status, &(argv[1]));
+  assert(err == 0);
 
-  if (napi_make_callback(env, NULL, ctx, callback, 2, argv, NULL) == napi_pending_exception) {
-    napi_value fatal_exception;
-    napi_get_and_clear_last_exception(env, &fatal_exception);
-    napi_fatal_exception(env, fatal_exception);
-  }
+  err = js_call_function(env, ctx, callback, 2, argv, NULL);
```

Comment:
This should do a microtask checkpoint assuming that `on_udx_send()` is an I/O callback.